### ISSUE TITLE
Support for Git tag "aliases"

### ISSUE THREAD

**Issue by fad3t** [mooltiverse/nyx#243]
Hello,

I'm wondering if the current version of Nyx would support the creation of Git tag aliases.
For example, a new release for version "2.4.7" would also create/update the "2.4" and "2" tags.

If it's not there yet, I would like to look into it and contribute a bit ;)

Fred

**Comment by flelli** [comment]
Hi @fad3t ,

this is not supported as of today but I see your point and it makes sense.

It would be helpful if you could help me with the semantics so they can be generalized, then I can implement it.

**Comment by fad3t** [comment]
I'm not aware of any official semantics for this.
What I usually see is what is described above; i.e. tag `3` would point to the latest release with major version "3" etc.

I was thinking we could have an additional attribute under the releaseType, for example `gitTagAliases`.
This value would be `none` by default (= no aliases) but would accept `major`, `minor` or `both` - so Nyx knows what extra tags to create.

Another option could to be have an `extraTags` attribute that accepts templates, as:
```
"mainline": {
  "gitTag": true,
  "gitExtraTags": [
    "{{version.major}}",
    "{{version.major}},{{version.minor}}"
  ]
  <snip>
}
```
I find the 2nd option more flexible, but also less straight-forward, both in terms of development but also usage.

**Comment by flelli** [comment]
Hi @fad3t ,

this feature is now available for preview in [2.5.0-alpha.2](https://github.com/mooltiverse/nyx/releases/tag/2.5.0-alpha.2) and will be GA in the upcoming release [2.5.0](https://github.com/mooltiverse/nyx/releases/tag/2.5.0).

You can now configure multiple [tag names](https://mooltiverse.github.io/nyx/guide/user/configuration-reference/release-types/#git-tag-names) upon tagging. The configuration option accepts a list of templates and by default only contains one template (`{{version}}`, also for backward compatibility) but you can extend the list as you wish.

You can also use the new `versionMajorNumber`, `versionMinorNumber`, `versionPatchNumber`, `versionPreReleaseIdentifier`, `versionBuildMetadata` [state attributes](https://mooltiverse.github.io/nyx/guide/user/state-reference/global-attributes/) in tag name templates so I guess this completely fullfills your requirement.

**Comment by fad3t** [comment]
Hi @flelli,

Thanks a lot for implementing this so fast!
I just gave it a try and in my case it's partly working.

Here's my configuration:
```json
{
  "preset": "simple",
  "changelog": {
    "path": "CHANGELOG.md",
    "sections": {
      ":sparkles: Features": "^feat$",
      ":bug: Bug Fixes": "^fix$",
      ":memo: Documentation": "^docs$",
      ":lipstick: Styles": "^style$",
      ":white_check_mark: Tests": "^test$",
      ":green_heart: Continuous Integration": "^ci$",
      ":wrench: Chores": "^chore$"
    }
  },
  "verbosity": "INFO",
  "scheme": "SEMVER",
  "initialVersion": "0.1.0",
  "releaseLenient": true,
  "git": {
    "remotes": {
      "origin": {
        "authenticationMethod": "USER_PASSWORD",
        "user": "nyx",
        "password": "{{#environmentVariable}}GITLAB_TOKEN{{/environmentVariable}}"
      }
    }
  },
  "releaseTypes": {
    "enabled": [
      "mainline",
      "candidate"
    ],
    "publicationServices": [
      "gitlab"
    ],
    "items": {
      "mainline": {
        "matchBranches": "main",
        "gitTag": "true",
        "gitTagMessage": "Release version {{version}}",
        "gitTagNames": [
          "latest",
          "{{version}}",
          "{{versionMajorNumber}}",
          "{{versionMajorNumber}}.{{versionMinorNumber}}"
        ],
        "gitPush": "true",
        "publish": "true",
        "description": "{{#fileContent}}CHANGELOG.md{{/fileContent}}",
        "collapseVersions": false
      },
      "candidate": {
        "matchBranches": "feat",
        "gitTag": "true",
        "gitTagMessage": "Release version {{version}}",
        "gitPush": "true",
        "publish": "false",
        "collapseVersions": true,
        "collapsedVersionQualifier": "{{branch}}"
      }
    }
  },
  "services": {
    "gitlab": {
      "type": "GITLAB",
      "options": {
        "BASE_URI": "{{#environmentVariable}}CI_API_V4_URL{{/environmentVariable}}",
        "AUTHENTICATION_TOKEN": "{{#environmentVariable}}GITLAB_TOKEN{{/environmentVariable}}",
        "REPOSITORY_NAME": "{{#environmentVariable}}CI_PROJECT_NAME{{/environmentVariable}}",
        "REPOSITORY_OWNER": "{{#environmentVariable}}CI_PROJECT_NAMESPACE{{/environmentVariable}}"
      }
    }
  }
}
```
The first commit was fine, it created the 4 tags as expected. Then I committed another change of type "fix" meaning it should recreate 3 tags. However it fails with the following error: `an error occurred when trying to push: object not found`.

Some more context:
```
time="2023-08-20T19:39:21Z" level=debug msg="command 'MARK' is not up to date, running..."
time="2023-08-20T19:39:21Z" level=debug msg="running the Mark command..."
time="2023-08-20T19:39:21Z" level=debug msg="the release type has the git commit flag disabled"
time="2023-08-20T19:39:21Z" level=debug msg="the release type has the git tag flag enabled"
time="2023-08-20T19:39:21Z" level=debug msg="repository latest commit in HEAD branch is '4d2c904120fe0df57e42eb0fb42fe0c2370bfc55'"
time="2023-08-20T19:39:21Z" level=debug msg="tagging latest commit '4d2c904120fe0df57e42eb0fb42fe0c2370bfc55' with tag 'latest'"
time="2023-08-20T19:39:21Z" level=debug msg="the repository already had a tag 'latest' so it will be deleted first"
time="2023-08-20T19:39:21Z" level=debug msg="tagging as 'latest'"
time="2023-08-20T19:39:21Z" level=debug msg="repository latest commit in HEAD branch is '4d2c904120fe0df57e42eb0fb42fe0c2370bfc55'"
time="2023-08-20T19:39:21Z" level=debug msg="tag 'latest' applied to commit '4d2c904120fe0df57e42eb0fb42fe0c2370bfc55'"
time="2023-08-20T19:39:21Z" level=debug msg="tagging latest commit '4d2c904120fe0df57e42eb0fb42fe0c2370bfc55' with tag '0.10.5'"
time="2023-08-20T19:39:21Z" level=debug msg="tagging as '0.10.5'"
time="2023-08-20T19:39:21Z" level=debug msg="repository latest commit in HEAD branch is '4d2c904120fe0df57e42eb0fb42fe0c2370bfc55'"
time="2023-08-20T19:39:21Z" level=debug msg="tag '0.10.5' applied to commit '4d2c904120fe0df57e42eb0fb42fe0c2370bfc55'"
time="2023-08-20T19:39:21Z" level=debug msg="tagging latest commit '4d2c904120fe0df57e42eb0fb42fe0c2370bfc55' with tag '0'"
time="2023-08-20T19:39:21Z" level=debug msg="the repository already had a tag '0' so it will be deleted first"
time="2023-08-20T19:39:21Z" level=debug msg="tagging as '0'"
time="2023-08-20T19:39:21Z" level=debug msg="repository latest commit in HEAD branch is '4d2c904120fe0df57e42eb0fb42fe0c2370bfc55'"
time="2023-08-20T19:39:21Z" level=debug msg="tag '0' applied to commit '4d2c904120fe0df57e42eb0fb42fe0c2370bfc55'"
time="2023-08-20T19:39:21Z" level=debug msg="tagging latest commit '4d2c904120fe0df57e42eb0fb42fe0c2370bfc55' with tag '0.10'"
time="2023-08-20T19:39:21Z" level=debug msg="the repository already had a tag '0.10' so it will be deleted first"
time="2023-08-20T19:39:21Z" level=debug msg="tagging as '0.10'"
time="2023-08-20T19:39:21Z" level=debug msg="repository latest commit in HEAD branch is '4d2c904120fe0df57e42eb0fb42fe0c2370bfc55'"
time="2023-08-20T19:39:21Z" level=debug msg="tag '0.10' applied to commit '4d2c904120fe0df57e42eb0fb42fe0c2370bfc55'"
time="2023-08-20T19:39:21Z" level=debug msg="the release type has the git push flag enabled"
time="2023-08-20T19:39:21Z" level=debug msg="pushing local changes to remotes"
time="2023-08-20T19:39:21Z" level=debug msg="the list of remotes is not defined. Using the default remote 'origin'"
time="2023-08-20T19:39:21Z" level=debug msg="pushing local changes to remote 'origin'"
time="2023-08-20T19:39:21Z" level=debug msg="looking up credentials for remote 'origin'"
time="2023-08-20T19:39:21Z" level=debug msg="using configured credentials for remote 'origin'"
time="2023-08-20T19:39:21Z" level=debug msg="attempting push to 'origin' using user name and password credentials."
time="2023-08-20T19:39:21Z" level=debug msg="pushing changes to remote repository 'origin' using username and password"
time="2023-08-20T19:39:21Z" level=debug msg="username and password authentication will use custom authentication options"
an error occurred when trying to push: object not found
```
Note that removing the "duplicate" tags manually before running the job fixes the issue, so it probably has something to do with the delete/create tag part.
The service account used by Nyx is owner of the repository so I don't think this is about permissions issue.

**Comment by flelli** [comment]
Hi @fad3t ,

I'll look into this.

But first can I ask you to retry? I've experienced multiple issues with GitHub during the last couple of days and then they gradually fixed them all. Just to make sure this issue wasn't due to some temporary GitHub problem.

Thanks

**Comment by fad3t** [comment]
Hi @flelli,

We're using a self-hosted instance of GitLab, so I don't think this is related.
I just tried and behavior is the same.

Kr,
Fred

**Comment by fad3t** [comment]
Might be also worth noting we're running those commands in the CI, to avoid running in detached HEAD:
```
  - git config user.name "${GITLAB_USER_NAME}"
  - git config user.email "${GITLAB_USER_EMAIL}"
  - git checkout -b "${CI_COMMIT_REF_NAME}"
  - nyx --verbosity=DEBUG publish
```

**Comment by flelli** [comment]
Sorry, I missed you are running on GitLab.

So although the `object not found` message is not giving much of a hint, given that the first push worked and the second resulted in an error I guess it's about overwriting existing tags so I'll build an update with the `force` flag enabled when pushing.

Meanwhile, can you please check the tags you're using are not [protected](https://docs.gitlab.com/ee/user/project/protected_tags.html)?

**Comment by fad3t** [comment]
I checked, I'm not using protected tags at all.

**Comment by flelli** [comment]
Hi @fad3t ,

I spent a few hours trying to reproduce the issue but no success.

Nonetheless I published [2.5.0-alpha.2](https://github.com/mooltiverse/nyx/releases/tag/2.5.0-alpha.2), which forces push. Do you mind giving it a try?

Honestly, after these tests, I'm no longer confident it will solve the problem, but it's worth a try. Moreover I can't keep it in the final release because it may interfere with some protection rules.

Meanwhile I was wondering about other possible causes. For example, do you have any hooks in your repo?

I would also run another test just to collect more detailed output. The idea is to disable Nyx's push (se "gitPush": "false" in your configuration) and, instead, try running a `git push` with the `--verbose` flag from the command line to see if it gives any further output.

**Comment by fad3t** [comment]
Hi @flelli,

I tried with alpha.2 and it works perfectly.
I checked again the settings of the repository and they look OK; no protected branches/tags, removing tags with a `git push` is allowed, ..

I'll try the local `git push` tomorrow and keep you posted. Thanks again!

**Comment by fad3t** [comment]
I've cloned the repository and executed nyx (2.5.0-alpha.2) locally, with `gitPush` (and `gitPublish`) disabled.
Below the execution logs, nothing special to say:
```
<snip>
DEBU[0000] command 'MARK' is not up to date, running...
DEBU[0000] running the Mark command...
DEBU[0000] the release type has the git commit flag disabled
DEBU[0000] the release type has the git tag flag enabled
DEBU[0000] repository latest commit in HEAD branch is 'e760bb09aaf2a6409eb302306dbd2eb6d50ab3eb'
DEBU[0000] tagging latest commit 'e760bb09aaf2a6409eb302306dbd2eb6d50ab3eb' with tag 'latest'
DEBU[0000] the repository already had a tag 'latest' so it will be deleted first
DEBU[0000] tagging as 'latest'
DEBU[0000] repository latest commit in HEAD branch is 'e760bb09aaf2a6409eb302306dbd2eb6d50ab3eb'
DEBU[0000] tag 'latest' applied to commit 'e760bb09aaf2a6409eb302306dbd2eb6d50ab3eb'
DEBU[0000] tagging latest commit 'e760bb09aaf2a6409eb302306dbd2eb6d50ab3eb' with tag '0.11.2'
DEBU[0000] tagging as '0.11.2'
DEBU[0000] repository latest commit in HEAD branch is 'e760bb09aaf2a6409eb302306dbd2eb6d50ab3eb'
DEBU[0000] tag '0.11.2' applied to commit 'e760bb09aaf2a6409eb302306dbd2eb6d50ab3eb'
DEBU[0000] tagging latest commit 'e760bb09aaf2a6409eb302306dbd2eb6d50ab3eb' with tag '0'
DEBU[0000] the repository already had a tag '0' so it will be deleted first
DEBU[0000] tagging as '0'
DEBU[0000] repository latest commit in HEAD branch is 'e760bb09aaf2a6409eb302306dbd2eb6d50ab3eb'
DEBU[0000] tag '0' applied to commit 'e760bb09aaf2a6409eb302306dbd2eb6d50ab3eb'
DEBU[0000] tagging latest commit 'e760bb09aaf2a6409eb302306dbd2eb6d50ab3eb' with tag '0.11'
DEBU[0000] the repository already had a tag '0.11' so it will be deleted first
DEBU[0000] tagging as '0.11'
DEBU[0000] repository latest commit in HEAD branch is 'e760bb09aaf2a6409eb302306dbd2eb6d50ab3eb'
DEBU[0000] tag '0.11' applied to commit 'e760bb09aaf2a6409eb302306dbd2eb6d50ab3eb'
DEBU[0000] the release type has the git push flag disabled
DEBU[0000] storing the Mark command internal attributes to the State
DEBU[0000] repository latest commit in HEAD branch is 'e760bb09aaf2a6409eb302306dbd2eb6d50ab3eb'
DEBU[0000] command 'MARK' finished.
DEBU[0000] running command 'PUBLISH'
DEBU[0000] looking up command instance 'PUBLISH' from cache
DEBU[0000] no command instance 'PUBLISH' found in cache
DEBU[0000] creating new command instance 'PUBLISH'
DEBU[0000] new Publish command object
DEBU[0000] checking whether the Publish command is up to date
DEBU[0000] the Publish command is not up to date because the internal state has no version yet or the state version doesn't match the version previously published by Publish
DEBU[0000] command 'PUBLISH' is not up to date, running...
DEBU[0000] the release type has the publish flag disabled
DEBU[0000] storing the Publish command internal attributes to the State
DEBU[0000] command 'PUBLISH' finished.
```
However running `git tag -l` (or `git push`) produces the following error:
```
fatal: unexpected line in .git/packed-refs: ^aa170f6e2a9a82b3bfc1e00775aacadd668979cf
```
Here's the content of the file:
```
# pack-refs with: peeled fully-peeled sorted
17bb458c92ba6c99cefaeb69b2aef1f3338a37f6 refs/remotes/origin/dev-01
aa170f6e2a9a82b3bfc1e00775aacadd668979cf refs/remotes/origin/main
952e868af260707a44fae3b163c0933c228d9445 refs/remotes/origin/renovate/configure
^aa170f6e2a9a82b3bfc1e00775aacadd668979cf
4ddf3820a6dfda5aef0889d812aa912297ba122c refs/tags/0.10
^02d6f1813b96a2b0c6d6b2b20e93b697645e4d41
<snip>
```
It looks like the tag reference is missing, it should be:
```
# pack-refs with: peeled fully-peeled sorted
17bb458c92ba6c99cefaeb69b2aef1f3338a37f6 refs/remotes/origin/dev-01
aa170f6e2a9a82b3bfc1e00775aacadd668979cf refs/remotes/origin/main
952e868af260707a44fae3b163c0933c228d9445 refs/remotes/origin/renovate/configure
71acbc92aaa6c16ec81a96cab18fc912062ca156 refs/tags/0
^aa170f6e2a9a82b3bfc1e00775aacadd668979cf
4ddf3820a6dfda5aef0889d812aa912297ba122c refs/tags/0.10
^02d6f1813b96a2b0c6d6b2b20e93b697645e4d41
```
But I'm a bit confused because I'm running `2.5.0-alpha.2` - which "force" pushes to GitLab.
I must be doing something wrong, any idea?

**Comment by flelli** [comment]
Dear @fad3t , it looks like we've hit a brain wrecker!

I honestly have no idea about the `fatal: unexpected line in .git/packed-refs` error and there's nothing helpful on the internet.

It would be interesting to run the same test on a CI server, where the original error happened.

**Comment by flelli** [comment]
Anyway @fad3t ,

I'm going to make the `force` flag configurable, so you have it available but it won't break anything for other users who then have control on that. The default behavior will be like `force` is disabled, also for backward compatibility.

**Comment by fad3t** [comment]
I tried on another clean repository and I'm not getting this packed-refs error, so might be unrelated.
On that repository I'm getting the expected error during the `git push --verbose`:
```
$ git push --tags --verbose
Pushing to gitlab.example.net:dummy/test.git
To gitlab.example.net:dummy/test.git
 = [up to date]      0.1.1 -> 0.1.1
 = [up to date]      0.1.2 -> 0.1.2
 = [up to date]      0.1.3 -> 0.1.3
 = [up to date]      0.1.4 -> 0.1.4
 ! [rejected]        0 -> 0 (already exists)
 ! [rejected]        0.1 -> 0.1 (already exists)
 ! [rejected]        latest -> latest (already exists)
error: failed to push some refs to 'gitlab.example.net:dummy/test.git'
hint: Updates were rejected because the tag already exists in the remote.
```
Adding the `--force` flag solves the issue.

If I'm deleting the remote tags (`git push -d origin <TAG>`) before doing the push, then it's OK -- I think this is what Nyx is doing. I'm going to redo the CI test on another repository, but I don't have much hope.

But like you said, having a `force` flag should do, that's probably the easiest option.

**Comment by flelli** [comment]
Alright, I'm trying to publish a new alpha release asap with the `force` flag

**Comment by flelli** [comment]
Hi @fad3t  ,

these features are on their way. They will be available in preview in [2.5.0-alpha.4](https://github.com/mooltiverse/nyx/releases/tag/2.5.0-alpha.4) when the pipeline for [this PR](https://github.com/mooltiverse/nyx/pull/257) will complete.

Sorry to write you without waiting for the pipelines to run and complete but this way I can just upload work that I can control while I'm afk during the weekend, hoping I can ship everything for you to try asap in the weekend.

They will also be generally available in the upcoming [2.5.0](https://github.com/mooltiverse/nyx/releases/tag/2.5.0).

Two new configuration options are available in the release type configuration: [`gitPushForce`](https://mooltiverse.github.io/nyx/guide/user/configuration-reference/release-types/#git-push-force), [`gitTagForce`](https://mooltiverse.github.io/nyx/guide/user/configuration-reference/release-types/#git-tag-force) (links to the docs will work once the 2.5.0 release is available). Set them both to `true` and everything should work as you tested.

**Comment by fad3t** [comment]
Thanks @flelli, great stuff as usual!
FYI I tested alpha.2 on another blank repository and it worked, so I guess my original test repository was somehow corrupted..
